File: drills/ts/harnesses/nc76_kth_largest_stream.ts
import * as solution from "./solution";

const show = (v: unknown) => JSON.stringify(v) ?? "undefined";

const stream = (k: number, initial: number[], added: number[]) => {
  const store = new solution.KthLargest(k, initial);
  return added.map((value) => store.add(value));
};

export function run(): [string, string, string][] {
  if (typeof solution.KthLargest !== "function") throw new Error("__signature_mismatch__");
  return [
    ["k = 3 over [4, 5, 8, 2] then 3, 5, 10, 9, 4", show([4, 5, 5, 8, 8]), show(stream(3, [4, 5, 8, 2], [3, 5, 10, 9, 4]))],
    ["k = 1 over [] then 1, 2, 0", show([1, 2, 2]), show(stream(1, [], [1, 2, 0]))],
    ["k = 2 over [] then 5, 5", show([null, 5]), show(stream(2, [], [5, 5]))],
  ];
}

File: drills/go/harnesses/nc76_kth_largest_stream.go
package main

func stream(k int, initial []int, added []int) []int {
	store := Constructor(k, initial)
	out := []int{}
	for _, n := range added {
		out = append(out, store.Add(n))
	}
	return out
}

func main() {
	run(func() []testCase {
		return []testCase{
			tc("k = 3 over [4, 5, 8, 2] then 3, 5, 10, 9, 4", []int{4, 5, 5, 8, 8}, stream(3, []int{4, 5, 8, 2}, []int{3, 5, 10, 9, 4})),
			tc("k = 1 over [] then 1, 2, 0", []int{1, 2, 2}, stream(1, []int{}, []int{1, 2, 0})),
			tc("k = 2 over [7] then 5, 5", []int{5, 5}, stream(2, []int{7}, []int{5, 5})),
		}
	})
}
